### PR TITLE
[Bromley] Disallow opening if existing missed/request report

### PR DESCRIPTION
**Pull request by dracos**
Fixes https://github.com/mysociety/fixmystreet-commercial/issues/1909 [skip changelog]

### CODE REVIEW

File: perllib/FixMyStreet/Cobrand/Bromley.pm [hunk 1]
```diff
@@ -621,7 +618,7 @@ working days (excluding weekends and bank holidays) after that date.
 
 =cut
 
-sub report_allowed {
+sub report_allowed_time {
```
Comment (struan) [inline perllib/FixMyStreet/Cobrand/Bromley.pm:646]: I'm not sure this name is an improvement tbh. Not that I can think of a better one :|

File: perllib/FixMyStreet/Cobrand/Bromley.pm [hunk 2]
```diff
@@ -477,6 +477,28 @@ sub bin_services_for_address {
     my $result = $echo->GetServiceUnitsForObject($property->{uprn});
     return [] unless $result;
 
+    my $events = $echo->GetEventsForObject($property->{id});
+    my $open;
+    foreach (@$events) {
+        next if $_->{ResolvedDate} || $_->{ResolutionCodeId}; # XXX?
```
Comment (struan) [inline perllib/FixMyStreet/Cobrand/Bromley.pm:483]: is this comment still relevant? If so it's not very helpful :)

File: perllib/FixMyStreet/Cobrand/Bromley.pm [hunk 3]
```diff
@@ -477,6 +477,28 @@ sub bin_services_for_address {
...
+        my $event_type = $_->{EventTypeId};
+        my $service_id = $_->{ServiceId};
+        if ($event_type == 2104) { # Request
+            my $data = $_->{Data}{ExtensibleDatum};
+            my $container;
+            foreach (@$data) {
+                if ($_->{ChildData}) {
+                    foreach (@{$_->{ChildData}{ExtensibleDatum}}) {
+                        $container = $_->{Value} if $_->{DatatypeName} eq 'Container Type';
```
Comment (struan) [inline perllib/FixMyStreet/Cobrand/Bromley.pm:492]: does it make sense to have a `last` here once you match container? I assume there can be only one.

File: perllib/FixMyStreet/Cobrand/Bromley.pm [hunk 4]
```diff
@@ -477,6 +477,28 @@ sub bin_services_for_address {
...
+                    }
+                }
+            }
+            $open->{request}->{$container} = 1;
+        } elsif (2095 <= $event_type && $event_type <= 2103) { # Missed collection
+            $open->{missed}->{$service_id} = 1;
+        }
+    }
+
```
Comment (struan) [inline perllib/FixMyStreet/Cobrand/Bromley.pm:501]: It does feel like this function has got quite large now. Also, `$result` is miles away from where it's used now.

File: perllib/FixMyStreet/Cobrand/Bromley.pm (hunk 1 above)
Comment (dracos) [reply perllib/FixMyStreet/Cobrand/Bromley.pm:649]: within_two_working_days I guess describes what it does, the fact that gets assigned to "report_allowed" is immaterial

File: perllib/FixMyStreet/Cobrand/Bromley.pm (hunk 4 above)
Comment (dracos) [reply perllib/FixMyStreet/Cobrand/Bromley.pm:503]: yeah, ta, have reordered, and pulled out the schedule parsing and event parsing to their own functions.

**Pull request merged**